src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -146,6 +147,70 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
     dagRequest
   }
 
+  def addSortOrder(request: TiDAGRequest, sortOrder: Seq[SortOrder]): Unit =
+    if (sortOrder != null) {
+      sortOrder.foreach(
+        (order: SortOrder) =>
+          request.addOrderByItem(
+            TiByItem.create(
+              BasicExpression.convertToTiExpr(order.child).get,
+              order.direction.sql.equalsIgnoreCase("DESC")
+            )
+        )
+      )
+    }
+
+  def pruneTopNFilterProject(
+    limit: Int,
+    projectList: Seq[NamedExpression],
+    filterPredicates: Seq[Expression],
+    source: TiDBRelation,
+    sortOrder: Seq[SortOrder] = null
+  ): SparkPlan = {
+    val request = new TiDAGRequest(pushDownType())
+    request.setLimit(limit)
+    addSortOrder(request, sortOrder)
+    pruneFilterProject(projectList, filterPredicates, source, request)
+  }
+
+  def collectLimit(limit: Int, child: LogicalPlan): Seq[SparkPlan] =
+    takeOrderedAndProject(limit, null, child)
+
+  def takeOrderedAndProject(
+    limit: Int,
+    sortOrder: Seq[SortOrder],
+    child: LogicalPlan
+  ): Seq[SparkPlan] = {
+    // If sortOrder is not null, limit must be greater than 0
+    if (limit < 0 || (sortOrder == null && limit == 0)) {
+      return Nil
+    }
+
+    child match {
+      case PhysicalOperation(projectList, filters, LogicalRelation(source: TiDBRelation, _, _))
+          if filters.forall(TiUtils.isSupportedFilter(_, source)) =>
+        pruneTopNFilterProject(limit, projectList, filters, source) :: Nil
+      case TiAggregation(
```

Comment:
Would you please confirm if topN and aggregates can be pushdown independently. 
For example, if you have group by name and count
you have 
region 1: A, 100, B, 1
region 2: A, 100, B 1000
and you take Top 1. In above case, B will be ignored in each region request but actually B is the group to take.
I wonder if above case works in your implementation.